File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        
        
        def lca(node, target1, target2):
            if not node or node.val == target1.val or node.val == target2.val:
                return node
            if node.val > target1.val and node.val > target2.val:
                return lca(node.left, target1, target2)
            if node.val < target1.val and node.val < target2.val:
                return lca(node.right, target1, target2)
            return node
        
        return lca(root, p, q)

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return dfs(root, p, q);
    }
    
    
    TreeNode* dfs(TreeNode* node, TreeNode* p, TreeNode* q){
        if (node == NULL || node == p || node == q) return node;
        
        
        if (p->val < node->val && q->val < node->val) return dfs(node->left, p, q);
        
        if (p->val > node->val && q->val > node->val) return dfs(node->right, p, q);
            
        return node;
        
        
    }
};